pennylane/transforms/qcut.py

```diff
@@ -505,3 +506,320 @@ def contract_tensors(
+            Optional only when ``max_free_wires`` is provided.
+        max_free_wires (int): Number of wires for the largest available device. Optional only when
+            ``devices`` is provided where it defaults to the maximum number of wires among
+            ``devices``.
+        min_free_wires (int): Number of wires for the smallest available device.
+            Optional, defaults to ``max_device_wires``.
+        max_fragments_probed (Union[int, Sequence[int]]): Single, or 2-Sequence of, number(s)
+            specifying the potential (range of) number of fragments for the partitioner to attampt.
+            Optional, defaults to probing all valid strategies derivable from the circuit and
+            devices.
+        max_free_gates (int): Maximum allowed circuit depth for the deepest available device.
+            Optional, defaults to unlimited depth.
+        min_free_gates (int): Maximum allowed circuit depth for the shallowest available device.
+            Optional, defaults to ``max_device_gates``.
+
+    **Example**
+
+    .. code-block:: python
+
+        import pennylane as qml
+        from pennylane.transforms import qcut
+
+        dev_a = qml.device('default.qubit', wires=4)
+        dev_b = qml.device('default.qubit', wires=6)
+
+        with qml.tape.QuantumTape() as tape:
+            qml.RX(0.4, wires=0)
+            qml.RY(0.9, wires=0)
+            qml.CNOT(wires=[0, 1])
+            qml.expval(qml.PauliZ(1))
+
+        tape_dag = qcut.tape_to_graph(tape)
+
+    Cut circuit with single-device-based strategy using the default 'kahypar' cutter:
+    >>> cut_strategy = qcut.CutStrategy(devices=dev_a)
+    >>> qcut.cut_circuit(tape_dag, method='kahypar', strategy=cut_strategy)
+
+    Cut circuit with multi-device-based strategy using the default 'kahypar' cutter:
+    >>> cut_strategy = qcut.CutStrategy(devices=(dev_a, dev_b))
+    >>> qcut.cut_circuit(tape_dag, method='kahypar', strategy=cut_strategy)
+
+    Cut circuit with user-supplied strategy using user-supplied partitioning callable:
+    >>> cut_strategy = qcut.CutStrategy(
+            max_free_wires=6,
+            min_free_wires=4,
+            num_fragments_probed=(2, 5),
+        )
+    >>> qcut.cut_circuit(tape_dag, method=my_partitioner, strategy=cut_strategy, **my_kwargs)
+
+    """
+
+    # pylint: disable=too-many-arguments, too-many-instance-attributes
+
+    devices: InitVar[Sequence[qml.Device]] = None
+    max_free_wires: int = None
+    min_free_wires: int = None
+    num_fragments_probed: Union[int, Sequence[int]] = None
+    max_free_gates: int = None
+    min_free_gates: int = None
+    imbalance_tolerance: float = None
+
+    HIGH_NUM_FRAGMENTS: ClassVar[int] = 20
+    HIGH_PARTITION_ATTEMPS: ClassVar[int] = 20
+
+    def __post_init__(
+        self,
+        devices,
+    ):
+        """Deriving cutting constraints from given devices and parameters."""
+
+        self.max_free_wires = self.max_free_wires or self.min_free_wires
```

Comment:
This is me wanting to be slightly more forgiving before checking if one of `devices` and `max_free_wires` is provided, where if neither are provided but `min_free_wires` are, then I interpret `max_free_wires` to be the same as `min_free_wires`. Unlikely situation, but if it happens this interpretation should be safe.

Reply:
Thanks this makes sense, though now I'm not sure I understand the docstring where for `max_free_wires` we have "`Optional only when ``devices`` is provided` and vice versa for `devices`". Maybe I'm misunderstanding the use of `Optional` but this seems to suggest that at least one of the two must be passed?

Reply:
You are correct that the docstring doesn't accurately describe this behaviour. This was intentional since this is sort of a "failsafe" behaviour which doesn't need to be advertised to users since it adds nothing functionally new and could potentially confuse people even more. In the end, I do want to force users to provide either `devices` or `max_free_wires`.